File: backend/server_njs/controllers/userController.js
const userModel = require('../models/userModel');
const artistModel = require('../models/artistModel');
const albumModel = require('../models/albumModel');
const songModel = require('../models/songModel');
const loadController = require('./loadController')

class userController {
    constructor() { }

    async findUser(req, res) {
        try {
            const userId = req.params.id;
            const user = new userModel(null, null, null, null, null, null, null);
            const response = await user.getById(userId);
            res.status(200).json({ message: 'User found', results: response });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async updateUser(req, res) {
        try {
            const { id_user, firstName, lastName, email, password, profilePhoto } = req.body;
            const user = new userModel(id_user, firstName, lastName, email, password, null, profilePhoto);
            //Cambios para recibir base64 en la imagen de perfil
            if (profilePhoto != null) {
                const imageUrl = await loadController.uploadImage(profilePhoto);
                if (imageUrl) {
                    user.profilePhoto = imageUrl;
                }
            }
            
            const userUpdated = await user.update(user);
            if (userUpdated) {
                res.status(200).json({ message: 'User updated' });
            } else {
                res.status(500).json({ message: 'An error has occurred while uploading the User' });
            }
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async search(req, res) {
        try {
            const search = req.body.search;
            const artist = new artistModel(null, null, null, null);
            const album = new albumModel(null, null, null, null, null);
            const song = new songModel(null, null, null, null, null, null, null);
            const songs = await song.getByRegex(search);
            const albums = await album.getByRegex(search);
            const artists = await artist.getByRegex(search);
            res.status(200).json({ songs: songs, albums: albums, artists: artists });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async like(req, res) {
        try {
            const { userId, songId } = req.body;
            const user = new userModel(userId, null, null, null, null, null, null);
            const likedSong = await user.likeASong(songId);
            res.status(200).json({ message: 'Song liked by user' });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async unlike(req, res) {
        try {
            const { userId, songId } = req.body;
            const user = new userModel(userId, null, null, null, null, null, null);
            const unlikedSong = await user.unlikeASong(songId);
            res.status(200).json({ message: 'Song unliked by user' });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async randomSong(req, res) {
        try {
            const { userId } = req.body;
            const song = new songModel(null, null, null, null, null, null, null);
            const randomSong = await song.getRandom(userId);
            res.status(200).json({ song: randomSong });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async playSong(req, res) {
        try {
            const { userId, songId } = req.body;
            const song = new songModel(songId, null, null, null, null, null, null);
            const songObtained = await song.getToPlay(userId);
            res.status(200).json({ song: songObtained });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async getFavoriteSongs_By_User(req, res) {
        try {
            const { userId } = req.params;
            const user = new userModel(userId, null, null, null, null, null, null);
            const likedSongs = await user.getFavoriteSongs_By_User();
            res.status(200).json({ success: likedSongs });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async getTopSongs_By_User(req, res) {
        try {
            const { userId } = req.params;
            const user = new userModel(userId, null, null, null, null, null, null);
            const topSongs = await user.getTopSongs_By_User();
            res.status(200).json({ success: topSongs });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async getTopArtists_By_User(req, res) {
        try {
            const { userId } = req.params;
            const user = new userModel(userId, null, null, null, null, null, null);
            const topArtists = await user.getTopArtists_By_User();
            res.status(200).json({ success: topArtists });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async getTopAlbums_By_User(req, res) {
        try {
            const { userId } = req.params;
            const user = new userModel(userId, null, null, null, null, null, null);
            const topAlbums = await user.getTopAlbums_By_User();
            res.status(200).json({ success: topAlbums });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }

    async getHistorySongs_By_User(req, res) {
        try {
            const { userId } = req.params;
            const user = new userModel(userId, null, null, null, null, null, null);
            const historySongs = await user.getHistorySongs_By_User();
            res.status(200).json({ success: historySongs });
        } catch (err) {
            console.error(err);
            res.status(500).json({ message: 'Internal Server Error' });
        }
    }
}

module.exports = new userController();

File: backend/server_py/controllers/userController.py
from models.userModel import userModel
from models.artistModel import artistModel
from models.albumModel import albumModel
from models.songModel import songModel
from flask import Blueprint, jsonify, request

user_route = Blueprint('user_route', __name__)

@user_route.route('/play', methods=['POST'])
def play_song():
    try:
        data = request.get_json()
        userId = data['userId']
        songId = data['songId']
        song = songModel(songId, None, None, None, None, None, None)
        song_obtained = song.get_to_play(userId)
        return jsonify({'song': song_obtained}), 200
    except Exception as e:
        print(e)
        return jsonify({'message': 'Internal Server Error'}), 500

@user_route.route('/random', methods=['POST'])
def random_song():
    try:
        userId = request.json.get('userId')
        song = songModel(None, None, None, None, None, None, None)
        random_song = song.get_random(userId)
        return jsonify({'song': random_song}), 200
    except Exception as e:
        print(e)
        return jsonify({'message': 'Internal Server Error'}), 500

@user_route.route('/search', methods=['POST'])
def search():
    try:
        search = request.json.get('search')
        artist = artistModel(None, None, None, None)
        album = albumModel(None, None, None, None, None)
        song = songModel(None, None, None, None, None, None, None)
        songs = song.getByRegex(search)
        albums = album.getByRegex(search)
        artists = artist.getByRegex(search)
        return jsonify({ 'songs': songs, 'albums': albums, 'artists': artists}), 200
    except Exception as e:
        print(e)
        return jsonify({ 'message': 'Internal Server Error' }), 500

@user_route.route('/like', methods=['POST'])
def like():
    try:
        data = request.get_json()
        userId = data['userId']
        songId = data['songId']
        user = userModel(userId, None, None, None, None, None, None)
        likedSong = user.likeASong(songId)
        return jsonify({ 'message': 'Song liked by user' }), 200
    except Exception as e:
        print(e)
        return jsonify({ 'message': 'Internal Server Error' }), 500

@user_route.route('/unlike', methods=['POST'])
def unlike():
    try:
        data = request.get_json()
        userId = data['userId']
        songId = data['songId']
        user = userModel(userId, None, None, None, None, None, None)
        unlikedSong = user.unlikeASong(songId)
        return jsonify({ 'message': 'Song unliked by user' }), 200
    except Exception as e:
        print(e)
        return jsonify({ 'message': 'Internal Server Error' }), 500

@user_route.route('/favoriteSongs/<userId>', methods=['GET'])
def favoriteSongs(userId):
    try:
        user = userModel(userId, None, None, None, None, None, None)
        favoriteSongs = user.getFavoriteSongs_By_User()
        return jsonify({ 'success': favoriteSongs }), 200
    except Exception as e:
        print(e)
        return jsonify({ 'message': 'Internal Server Error' }), 500

@user_route.route('/topSongs/<userId>', methods=['GET'])
def topSongs(userId):
    try:
        user = userModel(userId, None, None, None, None, None, None)
        topSongs = user.getTopSongs_By_User()
        return jsonify({ 'success': topSongs }), 200
    except Exception as e:
        print(e)
        return jsonify({ 'message': 'Internal Server Error' }), 500
    
@user_route.route('/topArtists/<userId>', methods=['GET'])
def topArtists(userId):
    try:
        user = userModel(userId, None, None, None, None, None, None)
        topArtists = user.getTopArtists_By_User()
        return jsonify({ 'success': topArtists }), 200
    except Exception as e:
        print(e)
        return jsonify({ 'message': 'Internal Server Error' }), 500
    
@user_route.route('/topAlbums/<userId>', methods=['GET'])
def topAlbums(userId):
    try:
        user = userModel(userId, None, None, None, None, None, None)
        topAlbums = user.getTopAlbums_By_User()
        return jsonify({ 'success': topAlbums }), 200
    except Exception as e:
        print(e)
        return jsonify({ 'message': 'Internal Server Error' }), 500
    
@user_route.route('/history/<userId>', methods=['GET'])
def history(userId):
    try:
        user = userModel(userId, None, None, None, None, None, None)
        history = user.getHistorySongs_By_User()
        return jsonify({ 'success': history }), 200
    except Exception as e:
        print(e)
        return jsonify({ 'message': 'Internal Server Error' }), 500

